### PR TITLE
Full JS bindings for MaterialX Core

### PR DESCRIPTION
**Pull request by sdunkel**
This PR completes the set of bindings for the MaterialXCore package. Additional tests will follow.

Update: #1184

### CODE REVIEW

File: source/JsMaterialX/JsMaterialXCore/JsElement.cpp [hunk 1]
```diff
@@ -180,22 +192,7 @@ extern "C"
...
-            BIND_VALUE_ELEMENT_FUNC_INSTANCE(matrix44, mx::Matrix44)
-            BIND_VALUE_ELEMENT_FUNC_INSTANCE(string, std::string)
-            BIND_VALUE_ELEMENT_FUNC_INSTANCE(integerarray, mx::IntVec)
-            BIND_VALUE_ELEMENT_FUNC_INSTANCE(booleanarray, mx::BoolVec)
-            BIND_VALUE_ELEMENT_FUNC_INSTANCE(floatarray, mx::FloatVec)
-            BIND_VALUE_ELEMENT_FUNC_INSTANCE(stringarray, mx::StringVec);
+            .class_property("UI_ADVANCED_ATTRIBUTE", &mx::ValueElement::UI_ADVANCED_ATTRIBUTE);
 
         ems::class_<mx::Token, ems::base<mx::ValueElement>>("Token")
             .smart_ptr_constructor("Token", &std::make_shared<mx::Token, mx::ElementPtr, const std::string &>)
```
Comment (sdunkel) [inline source/JsMaterialX/JsMaterialXCore/JsElement.cpp:198]: Can't comment on line. `isResolvedType` on StringResolver missing

File: source/JsMaterialX/test/basics.spec.js [hunk 2]
```diff
@@ -111,10 +111,16 @@ describe('Basics', () => {
 
     it('Matrices', () => {
         // Translation and scale
-        const trans = new mx.Matrix44().createTranslation(new mx.Vector3(1, 2, 3));
-        const scale = new mx.Matrix44().createScale(new mx.Vector3(2, 2, 2));
-        expect(trans.equals(new mx.Matrix44(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 2, 3, 1)));
-        expect(scale.equals(new mx.Matrix44(2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 1)));
+        const trans = mx.Matrix44.createTranslation(new mx.Vector3(1, 2, 3));
```
Comment (frericp) [inline source/JsMaterialX/test/basics.spec.js:114]: The tests seem to be a bit inconsistent. We don't use `new` on the matrices, but on the vectors. Furthermore, we use mix of `equals`, `equals.[].to.be.true` and `isEquivalent`.

Comment (frericp) [reply source/JsMaterialX/test/basics.spec.js:114]: Oh, just noticed that `equals` and `isEquivalent` are mx methods, not mocha constructs...

File: source/JsMaterialX/test/codeExamples.spec.js [hunk 3]
```diff
@@ -88,12 +88,12 @@ describe('Code Examples', () => {
 
         // Traverse the document tree in depth-first order.
         const elements = doc.traverseTree();
-        let elem = elements.next();
-        expect(elem).to.exist;
+        //let elem = elements.next();
+        //expect(elem).to.exist;
```
Comment (frericp) [inline source/JsMaterialX/test/codeExamples.spec.js:92]: Can we remove this?

File: source/JsMaterialX/test/codeExamples.spec.js [hunk 4]
```diff
@@ -106,7 +106,7 @@ describe('Code Examples', () => {
                     expect(filename).to.equal('greysphere_calibration.png');
                 }
             }
-            elem = elements.next();
+            //elem = elements.next();
```
Comment (frericp) [inline source/JsMaterialX/test/codeExamples.spec.js:109]: Same here

File: source/JsMaterialX/test/readXml.spec.js [hunk 5]
```diff
@@ -13,92 +13,72 @@ const _exampleFilenames = [
     'PreShaderComposite.mtlx',
 ];
 
-describe('Build Document', () => {
+it('Read XML', async () => {
```
Comment (frericp) [inline source/JsMaterialX/test/readXml.spec.js:16]: Is this now aligned with the python workflow tests?

**Review by frericp — COMMENTED** [review]
Looking good already. I think we should streamline some things in a clean-up pass, like consistent typedefs, macro interfaces (including documented usage, e.g. with respect to closing semicolons), and indentation.

**Comment by frericp** [comment]
How did you organize the bindings so far, regarding their order? Are they defined in the same order as in the C++ headers?

File: source/JsMaterialX/JsMaterialXCore/JsTraversal.cpp [hunk 6]
```diff
@@ -9,6 +9,18 @@
...
+#define BIND_ITERABLE_PROTOCOL(NAME)                                                       \
+            .function("next", ems::optional_override([](mx::NAME &it) {                    \
+                bool done = ++it == it.end();                                              \
+                ems::val result = ems::val::object();                                      \
+                result.set("done", done);                                                  \
+                if (!done) result.set("value", *it);                                       \
+                return result;                                                             \
+            }));                                                                           \
+            EM_ASM(                                                                        \
+                Module[#NAME]['prototype'][Symbol.iterator] = function() { return this; }; \
```
[the remark this answers is not in the document]
Comment (sdunkel) [reply source/JsMaterialX/JsMaterialXCore/JsTraversal.cpp:21]: It refers to the iterator. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols

File: source/JsMaterialX/JsMaterialXCore/JsTypes.cpp [hunk 7]
```diff
@@ -8,150 +8,135 @@
...
-        ems::class_<mx::Vector2, ems::base<mx::VectorBase>>("Vector2")
-            .constructor<>()
-            .constructor<float, float>()
-                BIND_VECTOR_SUBCLASS(mx::Vector2)
-            .function("cross", &mx::Vector2::cross);
+    ems::class_<mx::Vector2, ems::base<mx::VectorBase>>("Vector2")
+        .constructor<>()
+        .constructor<float, float>()
+            BIND_VECTOR_SUBCLASS(mx::Vector2)
+        .function("cross", &mx::Vector2::cross);
```
[the remark this answers is not in the document]
Comment (sdunkel) [reply source/JsMaterialX/JsMaterialXCore/JsTypes.cpp:61]: exactly

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp [hunk 8]
```diff
@@ -43,6 +43,7 @@ extern "C"
             .function("getLook", &mx::Document::getLook)
             .function("getLooks", &mx::Document::getLooks)
             .function("removeLook", &mx::Document::removeLook)
+            .function("mergeLooks", &mx::Document::mergeLooks)
```
[the remark this answers is not in the document]
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:46]: There is actually two functions, one is a static class function the other one a global one. The static class one is missing though. Good catch.

File: source/JsMaterialX/JsMaterialXCore/JsElement.cpp [hunk 9]
```diff
@@ -30,100 +32,92 @@ extern "C"
             .smart_ptr<std::shared_ptr<const mx::Element>>("Element") // mx::ConstElementPtr
             .function("setCategory", &mx::Element::setCategory)
             .function("getCategory", &mx::Element::getCategory)
-            .function("getName", &mx::Element::getName)
             .function("setName", &mx::Element::setName)
-            .function("getNamePath", &mx::Element::getNamePath) // might need to do something with the mx::ConstElementPtr relativeTo parameter
-            .function("setInheritsFrom", &mx::Element::setInheritsFrom)
-            .function("hasInheritedBase", &mx::Element::hasInheritedBase)
+            .function("getName", &mx::Element::getName)
+            BIND_MEMBER_FUNC_RAW_PTR("getNamePath", mx::Element, getNamePath, 0, 1, mx::ConstElementPtr)
```
[the remark this answers is not in the document]
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsElement.cpp:37]: That is what I thought, maybe it is not necessary as it is a `shared_ptr`? @frericp

File: source/JsMaterialX/JsMaterialXCore/JsLook.cpp [hunk 10]
```diff
@@ -79,6 +98,10 @@ extern "C"
             .function("getVisibilityType", &mx::Visibility::getVisibilityType)
             .function("setVisible", &mx::Visibility::setVisible)
             .function("getVisible", &mx::Visibility::getVisible)
-            .class_property("CATEGORY", &mx::Visibility::CATEGORY);
+            .class_property("CATEGORY", &mx::Visibility::CATEGORY)
+            .class_property("VIEWER_GEOM_ATTRIBUTE", &mx::Visibility::VIEWER_GEOM_ATTRIBUTE)
+            .class_property("VIEWER_COLLECTION_ATTRIBUTE", &mx::Visibility::VIEWER_COLLECTION_ATTRIBUTE)
+            .class_property("VISIBILITY_TYPE_ATTRIBUTE", &mx::Visibility::VISIBILITY_TYPE_ATTRIBUTE)
+            .class_property("VISIBLE_ATTRIBUTE", &mx::Visibility::VISIBLE_ATTRIBUTE);
     }
```
Comment (sdunkel) [inline source/JsMaterialX/JsMaterialXCore/JsLook.cpp:106]: getGeometryBindings missing

**Review by sdunkel — COMMENTED** [review]
we don't have the bindings for stuff in material.h

**Comment by sdunkel** [comment]
Yes, we try to have the same order in cpp and js binding to make it easier to compare both

File: source/JsMaterialX/JsMaterialXCore/JsElement.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (frericp) [reply source/JsMaterialX/JsMaterialXCore/JsElement.cpp:37]: You can use the normal `BIND_MEMBER_FUNC` macro if the parameter (or return value) is a smart pointer (for custom types, you probably need a `smart_ptr_constructor` on the class that the pointer refers to). The `RAW_PTR` macros are only required if a parameter (or return value) is a raw pointer (i.e. `TYPE *` notation).

File: documents/DeveloperGuide/CodeExamples.md [hunk 11]
```diff
@@ -93,8 +93,8 @@ Module().then((_module) => {
 
     // Create a simple shader interface.
     const simpleSrf = doc.addNodeDef('ND_simpleSrf', 'surfaceshader', 'simpleSrf');
-    const diffColor = simpleSrf.setInputValuecolor3('diffColor', new mx.Color3(1.0, 1.0, 1.0));
-    const specColor = simpleSrf.setInputValuecolor3('specColor', new mx.Color3(0.0, 0.0, 0.0));
+    const diffColor = simpleSrf.setInputValueColor3('diffColor', new mx.Color3(1.0, 1.0, 1.0));
```
Comment (frericp) [inline documents/DeveloperGuide/CodeExamples.md:96]: Please undo the changes in this file. This is the documentation for C++ :)
Btw, this made me notice that we're now diverging from the C++ interface, by renaming these functions...

Comment (muenstc) [reply documents/DeveloperGuide/CodeExamples.md:96]: Ah, I'm sorry :D

File: source/JsMaterialX/JsMaterialXCore/JsTypes.cpp [hunk 12]
```diff
@@ -83,75 +96,50 @@ extern "C"
...
-                     return mx::NAME_PATH_SEPARATOR;
-                 }));
-        ems::function("ARRAY_VALID_SEPARATORS", ems::optional_override([]() {
-                     return mx::ARRAY_VALID_SEPARATORS;
-                 }));
-        ems::function("ARRAY_PREFERRED_SEPARATOR", ems::optional_override([]() {
-                     return mx::ARRAY_PREFERRED_SEPARATOR;
-                 }));
+        ems::constant("DEFAULT_TYPE_STRING", mx::DEFAULT_TYPE_STRING);
+        ems::constant("EMPTY_STRING", mx::EMPTY_STRING);
```
Comment (muenstc) [inline source/JsMaterialX/JsMaterialXCore/JsTypes.cpp]: Why is the EMPTY_STRING defined here?

File: source/JsMaterialX/JsMaterialXCore/JsTypes.js [hunk 13]
```diff
@@ -7,20 +7,20 @@ addWrapper(function(Module, api) {
...
-    api.MULTI_OUTPUT_TYPE_STRING = Module.MULTI_OUTPUT_TYPE_STRING();
-    api.NONE_TYPE_STRING = Module.NONE_TYPE_STRING();
-    api.VALUE_STRING_TRUE = Module.VALUE_STRING_TRUE();
-    api.VALUE_STRING_FALSE = Module.VALUE_STRING_FALSE();
-    api.NAME_PREFIX_SEPARATOR = Module.NAME_PREFIX_SEPARATOR();
-    api.NAME_PATH_SEPARATOR = Module.NAME_PATH_SEPARATOR();
-    api.ARRAY_VALID_SEPARATORS = Module.ARRAY_VALID_SEPARATORS();
-    api.ARRAY_PREFERRED_SEPARATOR = Module.ARRAY_PREFERRED_SEPARATOR();
+    api.DEFAULT_TYPE_STRING = Module.DEFAULT_TYPE_STRING;
+    api.EMPTY_STRING = Module.EMPTY_STRING;
```
Comment (muenstc) [inline source/JsMaterialX/JsMaterialXCore/JsTypes.js:11]: same as above

File: source/JsMaterialX/JsMaterialXCore/JsUnit.cpp [hunk 14]
```diff
@@ -16,23 +16,28 @@ extern "C"
         ems::class_<mx::UnitConverter>("UnitConverter")
             .smart_ptr<std::shared_ptr<mx::UnitConverter>>("UnitConverter")
             .smart_ptr<std::shared_ptr<const mx::UnitConverter>>("UnitConverter")
-            .function("convert_float", ems::select_overload<float(float, const std::string&, const std::string&) const>(&mx::UnitConverter::convert))
-            .function("convert_vector2", ems::select_overload<mx::Vector2(const mx::Vector2&, const std::string&, const std::string&) const>(&mx::UnitConverter::convert))
-            .function("convert_vector3", ems::select_overload<mx::Vector3(const mx::Vector3&, const std::string&, const std::string&) const>(&mx::UnitConverter::convert))
-            .function("convert_vector4", ems::select_overload<mx::Vector4(const mx::Vector4&, const std::string&, const std::string&) const>(&mx::UnitConverter::convert))
+            .function("convertFloat", ems::select_overload<float(float, const std::string&, const std::string&) const>(&mx::UnitConverter::convert))
```
Comment (muenstc) [inline source/JsMaterialX/JsMaterialXCore/JsUnit.cpp:19]: Why not follow the pattern here with having a `define BIND_CONVERT ...` up front?

File: source/JsMaterialX/JsMaterialXCore/JsUnit.cpp [hunk 15]
```diff
@@ -16,23 +16,28 @@ extern "C"
...
 
         ems::class_<mx::LinearUnitConverter, ems::base<mx::UnitConverter>>("LinearUnitConverter")
             .smart_ptr<std::shared_ptr<mx::LinearUnitConverter>>("LinearUnitConverter")
             .smart_ptr<std::shared_ptr<const mx::LinearUnitConverter>>("LinearUnitConverter")
-            .function("convert_float", ems::select_overload<float(float, const std::string&, const std::string&) const>(&mx::LinearUnitConverter::convert))
-            .function("convert_vector2", ems::select_overload<mx::Vector2(const mx::Vector2&, const std::string&, const std::string&) const>(&mx::LinearUnitConverter::convert))
-            .function("convert_vector3", ems::select_overload<mx::Vector3(const mx::Vector3&, const std::string&, const std::string&) const>(&mx::LinearUnitConverter::convert))
-            .function("convert_vector4", ems::select_overload<mx::Vector4(const mx::Vector4&, const std::string&, const std::string&) const>(&mx::LinearUnitConverter::convert))
+            .function("convertFloat", ems::select_overload<float(float, const std::string&, const std::string&) const>(&mx::LinearUnitConverter::convert))
```
Comment (muenstc) [inline source/JsMaterialX/JsMaterialXCore/JsUnit.cpp:30]: Function order not following Unit.h declarations

File: source/JsMaterialX/JsMaterialXCore/JsUtil.cpp [hunk 16]
```diff
@@ -16,38 +8,27 @@ namespace mx = MaterialX;
 
 extern "C"
 {
-    EMSCRIPTEN_BINDINGS(my_module)
+    EMSCRIPTEN_BINDINGS(util)
```
Comment (muenstc) [inline source/JsMaterialX/JsMaterialXCore/JsUtil.cpp:11]: Missing binding: `isValidNamespace`
What about the `hashCombine` func?

File: source/JsMaterialX/JsMaterialXCore/JsValue.cpp [hunk 17]
```diff
@@ -8,38 +8,80 @@ namespace ems = emscripten;
...
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(Vector3, mx::Vector3)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(Vector4, mx::Vector4)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(Matrix33, mx::Matrix33)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(Matrix44, mx::Matrix44)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(String, std::string)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(IntegerArray, mx::IntVec)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(BooleanArray, mx::BoolVec)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(FloatArray, mx::FloatVec)
+            BIND_TYPE_SPECIFIC_VALUE_FUNCS(StringArray, mx::StringVec)
+            ;
```
Comment (muenstc) [inline source/JsMaterialX/JsMaterialXCore/JsValue.cpp:67]: extra newline

File: source/JsMaterialX/JsMaterialXCore/JsUnit.cpp (hunk 14 above)
Comment (sdunkel) [reply source/JsMaterialX/JsMaterialXCore/JsUnit.cpp:19]: I thought it not worth it for the two classes and few functions

File: source/JsMaterialX/JsMaterialXCore/JsUtil.cpp (hunk 16 above)
Comment (sdunkel) [reply source/JsMaterialX/JsMaterialXCore/JsUtil.cpp:11]: hashCombine is only used once internally and also not exposed by python so for now I would not expose it

File: source/JsMaterialX/test/basics.spec.js (hunk 2 above)
Comment (sdunkel) [reply source/JsMaterialX/test/basics.spec.js:114]: What is not consistent in regarding `new`? When creating Vectors AND Matrices directly via values we use the constructor. If we want to get special matrices like translations we use the specific creator functions

Comment (frericp) [reply source/JsMaterialX/test/basics.spec.js:114]: Yep, my bad. I only saw the `new` being removed in front of the matrices, but overlooked that the methods have been changed into static functions in the bindings.

File: source/JsMaterialX/JsMaterialXCore/JsDefinition.cpp [hunk 18]
```diff
@@ -21,16 +26,20 @@ extern "C"
             .function("setNodeGroup", &mx::NodeDef::setNodeGroup)
```
[the remark this answers is not in the document]
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsDefinition.cpp:28]: added

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp [hunk 19]
```diff
@@ -62,7 +63,14 @@ extern "C"
             .function("getNodeDefs", &mx::Document::getNodeDefs)
             .function("removeNodeDef", &mx::Document::removeNodeDef)
             .function("getMatchingNodeDefs", &mx::Document::getMatchingNodeDefs)
-            .function("getMatchingImplementations", &mx::Document::getMatchingImplementations)
```
[the remark this answers is not in the document]
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:65]: As discussed: will leave it for now but it is likely a bug and the type of the parameter should also be `const std::string&`

File: source/JsMaterialX/JsMaterialXCore/JsUtil.cpp [hunk 20]
```diff
@@ -16,38 +8,27 @@ namespace mx = MaterialX;
...
-                         std::string key = keys[i].as<std::string>().c_str();
-                         std::string value = newValue[key].as<std::string>();
-                         separatorMapper[key] = value;
-                     }
-                     return mx::replaceSubstrings(str, separatorMapper);
-                 }));
+        ems::function("isValidName", &mx::isValidName);
+        ems::function("incrementName", &mx::incrementName);
 
-        ems::function("prettyPrint", &mx::prettyPrint);
```
[the remark this answers is not in the document]
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsUtil.cpp:51]: added it to JsElement.cpp

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp [hunk 21]
```diff
@@ -17,14 +18,16 @@ extern "C"
         ems::class_<mx::Document, ems::base<mx::GraphElement>>("Document")
             .smart_ptr_constructor("Document", &std::make_shared<mx::Document, mx::ElementPtr, const std::string &>)
             .smart_ptr<std::shared_ptr<const mx::Document>>("Document")
+             // At the moment only the Document type is used. Once more types are added this binding needs to be adapted accordingly.
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:21]: @muenstc What does this comment mean?

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp [hunk 22]
```diff
@@ -83,12 +95,21 @@ extern "C"
             .function("getUnitTypeDef", &mx::Document::getUnitTypeDef)
             .function("getUnitTypeDefs", &mx::Document::getUnitTypeDefs)
             .function("removeUnitTypeDef", &mx::Document::removeUnitTypeDef)
+            .function("getVersionIntegers", ems::optional_override([](mx::Document &self) {
+                auto version = mx::getVersionIntegers();
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:99]: This is calling the wrong function. It should invoke `self.getVersionIntegers()`!

Comment (frericp) [reply source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:99]: And the pointer indices below need to be updated accordingly, since `Node::getVersionIntegers` only returns a pair of two ints. Ideally, this is one of the functions that we test explicitly.

File: source/JsMaterialX/JsMaterialXCore/JsGeom.cpp [hunk 23]
```diff
@@ -1,58 +1,72 @@
...
             .function("hasGeom", &mx::GeomElement::hasGeom)
             .function("getGeom", &mx::GeomElement::getGeom)
+            .function("getActiveGeom", &mx::GeomElement::getActiveGeom)
             .function("setCollectionString", &mx::GeomElement::setCollectionString)
             .function("hasCollectionString", &mx::GeomElement::hasCollectionString)
             .function("getCollectionString", &mx::GeomElement::getCollectionString)
             .function("setCollection", &mx::GeomElement::setCollection)
-            .function("getCollection", &mx::GeomElement::getCollection);
+            .function("getCollection", &mx::GeomElement::getCollection)
+            // BIND_VALIDATE(mx::GeomElement)
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsGeom.cpp:38]: Why is this commented?

File: source/JsMaterialX/JsMaterialXCore/JsGeom.cpp [hunk 24]
```diff
@@ -94,13 +110,12 @@ extern "C"
             .function("getIncludeCollections", &mx::Collection::getIncludeCollections)
             .function("hasIncludeCycle", &mx::Collection::hasIncludeCycle)
             .function("matchesGeomString", &mx::Collection::matchesGeomString)
-            .class_property("CATEGORY", &mx::Collection::CATEGORY);
+            // BIND_VALIDATE(mx::Collection)
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsGeom.cpp:113]: Same here

File: source/JsMaterialX/JsMaterialXCore/JsInterface.cpp [hunk 25]
```diff
@@ -51,69 +75,65 @@ extern "C"
...
             .function("hasTarget", &mx::InterfaceElement::hasTarget)
-
+            .function("getTarget", &mx::InterfaceElement::getTarget)
             .function("setVersionString", &mx::InterfaceElement::setVersionString)
-            .function("getVersionString", &mx::InterfaceElement::getVersionString)
             .function("hasVersionString", &mx::InterfaceElement::hasVersionString)
+            .function("getVersionString", &mx::InterfaceElement::getVersionString)
+            .function("setVersionIntegers", &mx::InterfaceElement::getVersionIntegers)
+            .function("getVersionIntegers", ems::optional_override([](mx::InterfaceElement &self) {
+                auto version = mx::getVersionIntegers();
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsInterface.cpp:130]: Same issue as in the `Node` bindings.

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp [hunk 26]
```diff
@@ -17,14 +18,16 @@ extern "C"
...
             .function("copy", &mx::Document::copy)
             .function("importLibrary", &mx::Document::importLibrary)
             .function("getReferencedSourceUris", ems::optional_override([](mx::Document &self) {
-                          mx::StringSet referenced = self.mx::Document::getReferencedSourceUris();
-                          int size = referenced.size();
-                          return arrayToVec((std::string *)&referenced, size);
-                      }))
+                mx::StringSet referenced = self.mx::Document::getReferencedSourceUris();
+                int size = referenced.size();
+                return arrayToVec((std::string *)&referenced, size);
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:29]: I wonder if we could save a copy by creating a `val::array` instead of a vector here, similar to how we do it in the `getVersionIntegers` wrapper? This is also a candidate for explicit testing.

File: source/JsMaterialX/JsMaterialXCore/JsMaterial.cpp [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+extern "C"
+{
+    EMSCRIPTEN_BINDINGS(material)
+    {
+        ems::function("getShaderNodes", ems::optional_override([](mx::NodePtr materialNode) {
+            std::unordered_set<mx::NodePtr> set =  mx::getShaderNodes(materialNode);
+            // Put all elements into an std::vector
+            std::vector<mx::NodePtr> vec;
+            vec.insert(vec.end(), set.begin(), set.end());
+            return vec;
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsMaterial.cpp:23]: This could be done with the `arrayToVec` helper. Note: Don't try to construct a `val::array` directly here, as it will likely not work with smart pointers. Same goes for the other bindings in this file.

File: source/JsMaterialX/JsMaterialXCore/JsTraversal.cpp [hunk 28]
```diff
@@ -45,26 +55,18 @@ extern "C"
...
-                return *it;
-            }));
+            BIND_ITERABLE_PROTOCOL(GraphIterator)
 
         ems::class_<mx::InheritanceIterator>("InheritanceIterator")
             .smart_ptr_constructor("InheritanceIterator", &std::make_shared<mx::InheritanceIterator, mx::ConstElementPtr>)
+            BIND_ITERABLE_PROTOCOL(InheritanceIterator)
+
+
+        // TODO Wrap ExceptionFoundCycle?
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsTraversal.cpp:65]: What's the conclusion on this TODO?

File: source/JsMaterialX/JsMaterialXCore/JsUtil.cpp [hunk 29]
```diff
@@ -1,53 +1,38 @@
...
         ems::function("getVersionString", &mx::getVersionString);
-        ems::function("createValidName", &mx::createValidName); // arg0 === {std::string}, arg1 === {unicode representing character}
-        ems::function("isValidName", &mx::isValidName);
-        ems::function("incrementName", &mx::incrementName);
 
-        // The following function throw: Cannot call {function name} due to unbound types: XXXXX
         ems::function("getVersionIntegers", ems::optional_override([]() {
-                     std::tuple<int, int, int> version = mx::getVersionIntegers();
-                     return arrayToVec((int *)&version, 3);
+                     auto version = mx::getVersionIntegers();
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsUtil.cpp:19]: I would still prefer to make the `std::tuple<int, int, int>` explicit, since the pointer arithmetic below depends on the number of elements. With `auto`, it's not obvious how many elements need to be dealt with, and more importantly, it won't fail at compilation time if the number of elements (i.e. return type of the function) ever changes. 
This goes for all `getVersionIntegers` and related functions where we apply this wrapping.

File: source/JsMaterialX/JsMaterialXCore/JsUtil.cpp [hunk 30]
```diff
@@ -1,53 +1,38 @@
...
+                     return ems::val::array((int *)&version, (int *)&version + 3);
                  }));
 
-        ems::function("splitString", ems::optional_override([](std::string str, std::string sep) {
-                     const std::string &str1 = str;
-                     const std::string &sep2 = sep;
-                     return mx::splitString(str1, sep2);
-                 }));
+        ems::function("createValidName", ems::optional_override([](std::string name) {
```
Comment (frericp) [inline source/JsMaterialX/JsMaterialXCore/JsUtil.cpp:23]: This function takes an optional parameter. Why not use `BIND_FUNC("createValidName", createValidName, std::string, char)` and remove the other override below?

File: source/JsMaterialX/helpers.h [hunk 31]
```diff
@@ -347,4 +346,17 @@ API(JSNAME, ems::optional_override([](SELF) { \
 #define BIND_0_0(...) \
 BIND_0(__VA_ARGS__)
 
+#define BIND_VALIDATE(CLASSNAME)                                                        \
+  .function("validate", ems::optional_override([](CLASSNAME &self) {                    \
+    return self.CLASSNAME::validate();                                                  \
```
Comment (frericp) [inline source/JsMaterialX/helpers.h:351]: Do we need to call `self.CLASSNAME::validate()`? Why not simply `self.validate()`?

File: source/JsMaterialX/helpers.h [hunk 32]
```diff
@@ -347,4 +346,17 @@ API(JSNAME, ems::optional_override([](SELF) { \
...
+  }))                                                                                   \
+  .function("validate", ems::optional_override([](CLASSNAME &self, ems::val message) {  \
+      std::string nativeMessage;                                                        \
+      bool handleMessage = message.typeOf().as<std::string>() == "object";              \
+      bool res = self.CLASSNAME::validate(handleMessage ? &nativeMessage : nullptr);    \
```
Comment (frericp) [inline source/JsMaterialX/helpers.h:356]: Same here

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp (hunk 21 above)
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:21]: Currently in the c++ code the only call to the static class function `createDocument` is using the `Document` type.
That is why there is only this one binding.
If we followed the usual pattern of the Naming for typed Js Bindings it would actually become `createDocumentDocument` which is rather weird.
The comment is just to indicate that this would need to be revised if usages of `createDocument` with other types pop up.

File: source/JsMaterialX/JsMaterialXCore/JsGeom.cpp (hunk 23 above)
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsGeom.cpp:38]: Enabling it will break the tests whenever `doc.validated()` is invoked for some not yet discovered reason. Good catch it was something we wanted to look at again.

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp (hunk 22 above)
Comment (muenstc) [reply source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:99]: I think we should make it explicit as you mentioned. Pointer arithmetic for a pair is also not needed.

File: source/JsMaterialX/JsMaterialXCore/JsUtil.cpp (hunk 30 above)
Comment (sdunkel) [reply source/JsMaterialX/JsMaterialXCore/JsUtil.cpp:23]: Because the optional param is a char and not a string and therefore emscripten expects you to provide a number and not something like `"b"`
This is not ideal as you can provide more characters, but the tradeoff for usability seems to be better this way

File: source/JsMaterialX/JsMaterialXCore/JsDocument.cpp (hunk 21 above)
Comment (frericp) [reply source/JsMaterialX/JsMaterialXCore/JsDocument.cpp:19]: I see. It's probably fine to only expose this for Document now, even though you can use it with other types in the C++ API, of course. I'd suggest to still call this `createDocument` for the Document type later, and only append the other types to the name.

File: source/JsMaterialX/JsMaterialXCore/JsGeom.cpp [hunk 33]
```diff
@@ -1,58 +1,69 @@
...
+
 #include <MaterialXCore/Geom.h>
 
-#include <emscripten.h>
 #include <emscripten/bind.h>
 
 namespace ems = emscripten;
 namespace mx = MaterialX;
 
 #define BIND_GEOMINFO_FUNC_INSTANCE(NAME, T) \
```
Comment (bernardkwok) [inline source/JsMaterialX/JsMaterialXCore/JsGeom.cpp:11]: As a note. "#define" aren't really used for the library and "using" is used instead. I'm not sure if it applies to hear bit a merge head-up.

**Review by bernardkwok — APPROVED** [review]
LGTM. Thanks.
The nitpick comments I've left for merge time.
